drivers/adc/adc_ite_it8xxx2.c

```diff
@@ -0,0 +1,339 @@
+{
+	/* bit 0 : adc module disable */
+	IT83XX_ADC_ADCCFG &= ~0x01;
+	/* disable adc interrupt */
+	irq_disable(DT_INST_IRQN(0));
+}
+
+static int check_buffer_size(const struct adc_sequence *sequence,
+			     uint8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(uint16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+				sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_it8xxx2_start_read(const struct device *dev,
+			      const struct adc_sequence *sequence)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	uint8_t channels = sequence->channels;
+	uint8_t channel_count = 0;
+	uint8_t idx = 0;
+	uint8_t bit;
+	int err;
+
+	if (!channels || channels & ~BIT_MASK(CHIP_ADC_COUNT)) {
+		LOG_ERR("Invalid selection of channels");
+		return -EINVAL;
+	}
+
+	if (!sequence->resolution) {
+		LOG_ERR("ADC resolution is not valid");
+		return -EINVAL;
+	}
+	LOG_DBG("Configure resolution=%d", sequence->resolution);
+
+	bit = find_lsb_set(channels);
+	while (bit != 0) {
+		idx = bit - 1;
+		channels &= ~BIT(idx);
+		bit = find_lsb_set(channels);
+		channel_count++;
+	}
+
+	err = check_buffer_size(sequence, channel_count);
+	if (err) {
+		return err;
+	}
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_it8xxx2_data *data =
+		CONTAINER_OF(ctx, struct adc_it8xxx2_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+
+	/* enable adc interrupt */
+	irq_enable(DT_INST_IRQN(0));
+
+	/* bit 0 : adc module enable */
+	IT83XX_ADC_ADCCFG |= 0x01;
+}
+
+static int adc_it8xxx2_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	int err;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	err = adc_it8xxx2_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_it8xxx2_data *data =
+		CONTAINER_OF(ctx, struct adc_it8xxx2_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_data_valid(enum chip_adc_channel adc_ch)
+{
+	return IT83XX_ADC_ADCDVSTS & BIT(adc_ch);
+}
+
+static void adc_it8xxx2_get_sample(const struct device *dev)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	uint8_t channels = data->ctx.sequence.channels;
+	uint8_t idx = 0;
+	uint8_t valid = 0;
+	uint8_t bit;
+
+	bit = find_lsb_set(channels);
+	while (bit != 0) {
+		idx = bit - 1;
+		if (adc_data_valid(idx)) {
+
+			*data->buffer = ((*adc_ctrl_regs[idx].adc_datm << 8) +
```

Comment:
What is 8?

Can this be a enum / #define?

Reply:
This comment is still outstanding

Reply:
The *data->buffer is 16-bit, here is to move raw data(MSB) to the left by 8-bit. Does this 8 need to define?

Reply:
OK I see.

Please use | instead of + so it is clearer. Also you can separate out your 8-bit access from your calculation:

unsigned int val = *adc_ctrl_regs[idx].adc_datm << 8 | *adc_ctrl_regs[idx].adc_datl;

*data->buffer++ = val * data->ctx.sequence.resolution;